plugins/net_plugin/net_plugin.cpp

```diff
@@ -2409,6 +2394,7 @@ namespace eosio {
    bool connection::process_next_message( uint32_t message_length ) {
       try {
          // if next message is a block we already have, exit early
+         auto ds = pending_message_buffer.create_datastream();
```

Comment:
What was the reason to relocate the construction of this datastream here?  The immediately preceding comment plus the next 15 lines seem to exist for the express purpose of bypassing the (presumably expensive) construction of the datastream.  If creating the datastream is so cheap, I question the presence of the peek_datastream and its usage here.  It seems like extra work creating two objects, then rereading the same data twice (the block number), only to throw it away the second time.

Reply:
Creating the datastream is cheap. Using the datastream is not. But you are correct that this could be better although I would assume the optimizer would do something similar. I've moved the datastream construction closer to use.